python/faery/frame_filter.py

```diff
@@ -1,10 +1,37 @@
 import collections.abc
+import pathlib
 import types
 import typing
 
 import numpy
+import numpy.typing
 
-from . import frame_stream, stream
+from . import enums, frame_stream, stream, timestamp
+from .colormaps._base import Color, parse_color
+
+if typing.TYPE_CHECKING:
+    from .types import image  # type: ignore
+else:
+    from .extension import image
+
+
+SPEED_UP_PRECISION: float = 1e-3
+
+
+def number_to_string(number: typing.Union[int, float], precision: float) -> str:
```

Comment:
Why can't we use format strings for this? Like `f"{number:.2f}"`. It seems like you're trying to avoid some floating point rounding nastiness

Reply:
I added this to show integers when possible.

The function is intended to format the "speed up" ratio that appears on videos (that is, the duration represented by one event frame divided by the time that the frame is shown on screen by the video player). I quite like the following display rules:
- Show the speed up directly when it is larger than one (for instance `x 10` for a time-lapse where each event frame covers 1/6 seconds in a video where each frame is shown for 1/60 seconds).
- Show the speed up as an inverse when it is smaller than one (for instance `x 1/10` for a slow-motion video where each event frame covers 1/600 seconds in a video where each frame is shown for 1/60 seconds). This is particularly useful when discussing very slow motion videos ("ten thousand times slower than real-time", represented by `x 1/10000`, is easier to parse than `x 0.0001`).
- Show integers when close to a round value (`x 1/10` instead of `x 1/10.00`) but keep the fractional digits where needed (`x 1/1.4`, not `x 1/1`).

Python's formatting function gives us three options that all are all (sometimes) problematic:
- Let Python pick the number of digits (`f"x {number}"`), which can cause problems when the number is close to but not quite an integer.
- Always show an integer (`f"x {number:.0f}"`), which can be misleading when the number is a small non-integer.
- Always show digits (`f"x {number:.2f}"`), which often shows redundant fractional digits.

Reply:
There are also https://docs.python.org/3/library/stdtypes.html#float.as_integer_ratio and https://docs.python.org/3/library/fractions.html#fractions.Fraction. I agree that the use cases are great, but I'm just wondering whether some of the code could be simplified. It seems like a problem that someone in the Python ecosystem has faced before :)

Reply:
That's a fair point! I was not aware of as_integer_ratio. I'll try to refactor this to simplify the code.